stl/src/mutex.cpp

```diff
@@ -77,26 +77,28 @@ void _Mtx_destroy(_Mtx_t mtx) { // destroy mutex
 }
 
 static _Thrd_result mtx_do_lock(_Mtx_t mtx, const _timespec64* target) { // lock mutex
+    // TRANSITION, ABI: the use of `const _timespec64*` is preserved for `_Mtx_timedlock`
+    const long current_thread_id = static_cast<long>(GetCurrentThreadId());
     if ((mtx->_Type & ~_Mtx_recursive) == _Mtx_plain) { // set the lock
-        if (mtx->_Thread_id != static_cast<long>(GetCurrentThreadId())) { // not current thread, do lock
+        if (mtx->_Thread_id != current_thread_id) { // not current thread, do lock

```

Comment:
It seems that this branch `if ((mtx->_Type & ~_Mtx_recursive) == _Mtx_plain)` is only used in the outdated `_Thrd_create`?

Reply:
The legacy burden looks very heavy...

Reply:
Can we define new versions of  `_Mtx_lock` and `_Mtx_trylock` for their real active users (`std::mutex`; `std::recursive_mutex`)?

Reply:
I think that it's safe to add a comment.

Adding variants of `_Mtx_lock` and `_Mtx_trylock` would need to be done in the import lib. I cannot emphasize strongly enough how difficult it is to mess with the STL's separately compiled code (specifically the interface between the headers and the separately compiled code; changes that are entirely on one side are less risky) - we need to do it to add new features, and it can be worth it to fix important bugs or improve performance measurably. However, for pure cleanups, it is almost never worth the effort and risk. (The XP/Vista cleanups were an important counterexample, where we were able to eradicate so much code it was worth it, and this eventually enabled `constexpr mutex`.)

This is not exactly a "no" but it is a plea for caution and thought before attempting to make such changes. We would be persuaded by measurable benchmark improvements.

Because `mtx_do_lock` is `static`, you could create a variant of it for use by `_Mtx_lock`, but I am not sure if that would help improve performance or clarity.

Reply:
Thanks! I learned a lot from these comments.

Reply:
You're welcome! I apologize for sounding grumpy in some of them - I was running on little sleep for entirely unrelated reasons. You've been doing careful work (don't worry about the `_CRTIMP2_PURE` thing, nobody knows about that one until they run into it) and your PRs have identified good things to clean up (like the `WAIT_ABANDONED` dead code here). It is true that some cleanups aren't worth the risk, but we do still try to continuously improve the codebase which is what keeps the STL a nice place to implement features in, instead of an ossifying mass of legacy code. This PR is definitely a good cleanup!